numpy/core/src/umath/ufunc_object.c

```diff
@@ -2067,12 +2067,16 @@ _get_coredim_sizes(PyUFuncObject *ufunc, PyArrayObject **op,
 }
 
 /*
- * Returns a new reference
+ * Returns a new reference to the ufunc identity.  Note that this identity
+ * is only a default identity value stored on the ufunc, since the invidiual
+ * ufunc loop (ArrayMethod) is queried for the actual identity.
+ *
  * TODO: store a reference in the ufunc object itself, rather than
  *       constructing one each time
  */
-static PyObject *
-_get_identity(PyUFuncObject *ufunc, npy_bool *reorderable) {
+NPY_NO_EXPORT PyObject *
+PyUFunc_GetIdentity(PyUFuncObject *ufunc, npy_bool *reorderable)
```

Comment:
Should this be `_GetLegacyIdentity`? If indeed only used for old-style ufuncs, perhaps the function can be moved to `legacy_array_method`?

Or is this also meant to be used for new ufuncs? If so, would it be better to split out the `reorderable` and `identity` parts here too? (and adjust `ufunc_get_identity` accordingly).

Reply:
Honestly, I still need to figure that out (thus no change yet).  The problem is, we still have `ufunc.identity` and maybe someone actually uses that?  But if we want downstream to actually use this (rather than always implement the an initial getter), then we should probably add it as public API and give it some clearly name like `GetUFuncDefaultIdentity`.

Reply:
Indeed, we need to support `.identity` since someone might well use that. Looking at this again, I now agree it is best to just keep the code as close as possible as it is here, since we need the code as is in `legacy_array_method` and this works fine for getting the `.identity` attribute. Really what tripped me up here was the name change, both the name itself and the impression from the `PyUFunc` start that it would (eventually) be public. So, I guess the first question is whether we even need to change the name at all? I guess the reason for that is that the function is now used inside `legacy_array_method`, which makes sense. And I guess I forgot that `PyUfunc` just indicates "used elsewhere in the numpy C code"; it does have an `NPY_NO_EXPORT` after all!
Anyway, bottom line is that I think all we really need is a clearer name, and I like your `PyUFunc_GetDefaultIdentity` (removing the extra `UFunc` you had in your comment).

We can then leave open for later a possible public API (where perhaps one might want to get the identity for a given loop or so).

Reply:
OK, renamed to `PyUFunc_GetDefaultIdentity`... maybe it would make sense to have a pattern of using NPY for private and PyArray style for public. But I suspect we break it in both ways too often.